4-WebApp-your-API/4-1-MyOrg/TodoListService/Controllers/TodoListController.cs

```diff
@@ -27,66 +31,161 @@ public TodoListController(IHttpContextAccessor contextAccessor)
+            AcceptedAppPermission = new string[] { "ToDoList.Read.All", "ToDoList.ReadWrite.All" }
+            )]
         public IEnumerable<Todo> Get()
         {
-            string owner = User.Identity.Name;
-            return TodoStore.Values.Where(x => x.Owner == owner);
+            if (HasDelegatedPermissions(new string[] { "ToDoList.Read", "ToDoList.ReadWrite" }))
+            {
+                return TodoStore.Values.Where(x => x.Owner == GetObjectIdClaim(User));
+            }
+            else if (HasApplicationPermissions(new string[] { "ToDoList.Read.All", "ToDoList.ReadWrite.All" }))
+            {
+                return TodoStore.Values;
+            }
+
+            throw new ApplicationException("It's impossible for you to be in this state in a normal situation.");
         }
 
         // GET: api/values
         [HttpGet("{id}", Name = "Get")]
-        [RequiredScope(new string[] { "ToDoList.Read", "ToDoList.Write" })]
+        [RequiredScopeOrAppPermission(
+            AcceptedScope = new string[] { "ToDoList.Read", "ToDoList.ReadWrite" },
+            AcceptedAppPermission = new string[] { "ToDoList.Read.All", "ToDoList.ReadWrite.All" })]
         public Todo Get(int id)
         {
-            return TodoStore.Values.FirstOrDefault(t => t.Id == id);
+            //if it only has delegated permissions
+            //then it will be t.id==id && x.Owner == owner
+            //if it has app permissions the it will return  t.id==id
+
+            if (HasDelegatedPermissions(new string[] { "ToDoList.Read", "ToDoList.ReadWrite" }))
+            {
+                return TodoStore.Values.FirstOrDefault(t => t.Id == id && t.Owner == GetObjectIdClaim(User));
+            }
+            else if (HasApplicationPermissions(new string[] { "ToDoList.Read.All", "ToDoList.ReadWrite.All" }))
+            {
+                return TodoStore.Values.FirstOrDefault(t => t.Id == id);
+            }
+
+            throw new ApplicationException("It's impossible for you to be in this state. STINKY HACKER");
         }
 
         [HttpDelete("{id}")]
-        [RequiredScope("ToDoList.Write")]
+        [RequiredScopeOrAppPermission(
+            AcceptedScope = new string[] { "ToDoList.ReadWrite" },
+            AcceptedAppPermission = new string[] { "ToDoList.ReadWrite.All" })]
         public void Delete(int id)
         {
-            TodoStore.Remove(id);
+
+            if (
+                (
+
+                HasDelegatedPermissions(new string[] { "ToDoList.ReadWrite" }) && TodoStore.Values.Any(x => x.Id == id && x.Owner == GetObjectIdClaim(User)))
+
+                ||
+
+                HasApplicationPermissions(new string[] { "ToDoList.ReadWrite.All" })
+                )
+            {
+                TodoStore.Remove(id);
+            }
+
+            else
+            {
+                throw new ApplicationException("No idea how you've got here");
```

Comment:
That's not an actionable message!
You'd want to explain the context and have something actionable (for the developer of the app? or a developer calling the web api)?

Reply:
Yes, removed all these